### DISCUSSION TITLE
Deployment issues with Ubuntu 22.04 LTS and FreeBSD

### DISCUSSION THREAD

**Discussion by Externalnet** [ngoduykhanh/PowerDNS-Admin#1702] [Q&A]
### PDA version

0.4.1

### Python version

3.0

### Steps to Reproduce

root@powerdns:/var/www/html/pdns# pip install -r requirements.txt
Usage: pip [options]

ERROR: Invalid requirement: --use-feature=no-binary-enable-wheel-cache lxml==4.9.0
pip: error: option --use-feature: invalid choice: 'no-binary-enable-wheel-cache' (choose from '2020-resolver', 'fast-deps', 'in-tree-build')

root@powerdns:/var/www/html/pdns# source ./flask/bin/activate
(flask) root@powerdns:/var/www/html/pdns# export FLASK_APP=powerdnsadmin/__init__.py
(flask) root@powerdns:/var/www/html/pdns# flask db upgrade
Error: While importing 'powerdnsadmin', an ImportError was raised:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/flask/cli.py", line 256, in locate_app
    __import__(module_name)
  File "/var/www/html/pdns/powerdnsadmin/__init__.py", line 4, in <module>
    from flask_mail import Mail
ModuleNotFoundError: No module named 'flask_mail'

Usage: flask [OPTIONS] COMMAND [ARGS]...
Try 'flask --help' for help.

### Expected Behavior

Able to install PowerDNS Admin by following instuction/documentation

### Observed Behavior

root@powerdns:/var/www/html/pdns# pip install -r requirements.txt
Usage: pip [options]

ERROR: Invalid requirement: --use-feature=no-binary-enable-wheel-cache lxml==4.9.0
pip: error: option --use-feature: invalid choice: 'no-binary-enable-wheel-cache' (choose from '2020-resolver', 'fast-deps', 'in-tree-build')

root@powerdns:/var/www/html/pdns# source ./flask/bin/activate
(flask) root@powerdns:/var/www/html/pdns# export FLASK_APP=powerdnsadmin/__init__.py
(flask) root@powerdns:/var/www/html/pdns# flask db upgrade
Error: While importing 'powerdnsadmin', an ImportError was raised:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/flask/cli.py", line 256, in locate_app
    __import__(module_name)
  File "/var/www/html/pdns/powerdnsadmin/__init__.py", line 4, in <module>
    from flask_mail import Mail
ModuleNotFoundError: No module named 'flask_mail'

Usage: flask [OPTIONS] COMMAND [ARGS]...
Try 'flask --help' for help.

**Comment by metrax** [comment]
Please link the instructions you are using. There seems to be something wrong.

**Comment by Externalnet** [comment]
I used this instruction but also other instructions fail: https://www.howtoforge.com/how-to-install-powerdns-on-ubuntu-22-04/

**Comment by metrax** [comment]
Please use the official instructions:
https://github.com/PowerDNS-Admin/PowerDNS-Admin/tree/master/docs/wiki/install

In 0.4.1 there is a dependency problem in some distributions.
If you run in to an error because of "pg_config", you need to install libpg-dev (Debian/Ubuntu)
If you run in to an error because of "cython", change the Version Number of "PyYAML" to 6.0.1 in requirements.txt

I was installing PDA today at Ubuntu 22.04 and ran into these problems.

**Comment by Externalnet** [comment]
Thanks for all the quick replies, I will try the official instructions and come back her with the results.

**Comment by sonnh-uit** [comment]
I met same error on CentOS7. I follow instruction wiki above. Have any one figure out method to fix, please share for me.

**Comment by Externalnet** [comment]
Ok I used the official wiki instuction to install PowerDNS-Admin on my favorite OS FreeBSD v13
The first step in the instuction is:

"sudo pkg install git python3 curl node12 yarn-node12"

result:

Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
pkg: No packages available to install matching 'node12' have been found in the repositories
pkg: No packages available to install matching 'yarn-node12' have been found in the repositories

So this instuction is also not valid, what do I do next?

**Comment by metrax** [comment]
@Pammetje i have no clue about FreeBSD but sounds like the howto was written for an older FreeBSD version. Please check if these packages are available with new versions, but i'm not sure if newer yarn versions are working.

@sonnh-uit please post a full description of your problem with the steps you've done and the output of the errors.

**Comment by Externalnet** [comment]
There is a node-16, node-18 and a yarn-node16, yarn-node18 package on FreeBSD but not a node-12 and yarn-node12  package (probably end of life?). So wich version should I install, version 16 of version 18 ?

The second line in the FreeBSD instuction also is not working:
pkg install libxml2 libxslt pkgconf py37-xmlsec py37-cffi py37-ldap

Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
pkg: No packages available to install matching 'py37-xmlsec' have been found in the repositories
pkg: No packages available to install matching 'py37-cffi' have been found in the repositories
pkg: No packages available to install matching 'py37-ldap' have been found in the repositories

If I change that to:

pkg install libxml2 libxslt pkgconf py39-xmlsec py39-cffi py39-ldap

I can succesfully install the packages

I can succesfully git clone/check out the source code

When I try to make the virtual Python environment I run into the following error:

root@powerdns:/opt/powerdns-admin # python3 -m venv /web/python --system-site-packages
root@powerdns:/opt/powerdns-admin # source /web/python/bin/activate
Badly placed ()'s.
root@powerdns:/opt/powerdns-admin # 

If I use:
source /web/python/bin/activate.csh
it is working fine

This line in the FreeBSD instuction:
pip3 install -r requirements.txt
Gives me the folloing error/output:

```
Collecting Authlib==1.2.0 (from -r requirements.txt (line 1))
  Downloading Authlib-1.2.0-py2.py3-none-any.whl (214 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 214.8/214.8 kB 12.1 MB/s eta 0:00:00
Collecting Flask-Assets==2.0 (from -r requirements.txt (line 2))
  Downloading Flask_Assets-2.0-py3-none-any.whl (8.5 kB)
Collecting Flask-Login==0.6.2 (from -r requirements.txt (line 3))
  Downloading Flask_Login-0.6.2-py3-none-any.whl (17 kB)
Collecting Flask-Mail==0.9.1 (from -r requirements.txt (line 4))
  Downloading Flask-Mail-0.9.1.tar.gz (45 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.7/45.7 kB 7.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting Flask-Migrate==2.5.3 (from -r requirements.txt (line 5))
  Downloading Flask_Migrate-2.5.3-py2.py3-none-any.whl (13 kB)
Collecting Flask-SQLAlchemy==2.5.1 (from -r requirements.txt (line 6))
  Downloading Flask_SQLAlchemy-2.5.1-py2.py3-none-any.whl (17 kB)
Collecting Flask-SSLify==0.1.5 (from -r requirements.txt (line 7))
  Downloading Flask-SSLify-0.1.5.tar.gz (3.0 kB)
  Preparing metadata (setup.py) ... done
Collecting Flask-SeaSurf==1.1.1 (from -r requirements.txt (line 8))
  Downloading Flask_SeaSurf-1.1.1-py3-none-any.whl (8.4 kB)
Collecting Flask-Session==0.4.0 (from -r requirements.txt (line 9))
  Downloading Flask_Session-0.4.0-py2.py3-none-any.whl (7.5 kB)
Collecting Flask==2.1.3 (from -r requirements.txt (line 10))
  Downloading Flask-2.1.3-py3-none-any.whl (95 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.6/95.6 kB 22.0 MB/s eta 0:00:00
Collecting Jinja2==3.1.2 (from -r requirements.txt (line 11))
  Downloading Jinja2-3.1.2-py3-none-any.whl (133 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.1/133.1 kB 30.8 MB/s eta 0:00:00
Collecting PyYAML==5.4 (from -r requirements.txt (line 12))
  Downloading PyYAML-5.4.tar.gz (174 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 174.8/174.8 kB 38.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [62 lines of output]
      /tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/config/setupcfg.py:293: _DeprecatedConfig: Deprecated config in `setup.cfg`
      !!
      
              ********************************************************************************
              The license_file parameter is deprecated, use license_files instead.
      
              This deprecation is overdue, please update your project and remove deprecated
              calls to avoid build errors in the future.
      
              See https://setuptools.pypa.io/en/latest/userguide/declarative_config.html for details.
              ********************************************************************************
      
      !!
        parsed = self.parsers.get(option_name, lambda x: x)(value)
      running egg_info
      writing lib3/PyYAML.egg-info/PKG-INFO
      writing dependency_links to lib3/PyYAML.egg-info/dependency_links.txt
      writing top-level names to lib3/PyYAML.egg-info/top_level.txt
      Traceback (most recent call last):
        File "/web/python/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/web/python/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/web/python/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 355, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 325, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 341, in run_setup
          exec(code, locals())
        File "<string>", line 271, in <module>
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 185, in setup
          return run_commands(dist)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
          dist.run_commands()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/command/egg_info.py", line 318, in run
          self.find_sources()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/command/egg_info.py", line 326, in find_sources
          mm.run()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/command/egg_info.py", line 548, in run
          self.add_defaults()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/command/egg_info.py", line 586, in add_defaults
          sdist.add_defaults(self)
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/command/sdist.py", line 113, in add_defaults
          super().add_defaults()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/command/sdist.py", line 251, in add_defaults
          self._add_defaults_ext()
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/command/sdist.py", line 336, in _add_defaults_ext
          self.filelist.extend(build_ext.get_source_files())
        File "<string>", line 201, in get_source_files
        File "/tmp/pip-build-env-m42fvglf/overlay/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 107, in __getattr__
          raise AttributeError(attr)
      AttributeError: cython_sources
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

```




I am stuck here (the above was with node and yarn version 18, I also get the same error with version 16)

**Comment by 662KDV** [comment]
> Please use the official instructions: https://github.com/PowerDNS-Admin/PowerDNS-Admin/tree/master/docs/wiki/install
> 
> In 0.4.1 there is a dependency problem in some distributions. If you run in to an error because of "pg_config", you need to install libpg-dev (Debian/Ubuntu) If you run in to an error because of "cython", change the Version Number of "PyYAML" to 6.0.1 in requirements.txt
> 
> I was installing PDA today at Ubuntu 22.04 and ran into these problems.

I was seeing the same error as OP in Ubuntu 22.04
I can confirm the "cython" fix solved the error for me.

**Comment by sirEgghead** [comment]
I am on Ubuntu 22.04.  I began with OP's problem using the same guide as OP.

I changed to using https://github.com/PowerDNS-Admin/PowerDNS-Admin/blob/master/docs/wiki/install/Running-PowerDNS-Admin-on-Ubuntu-or-Debian.md when pointed to it by a comment above and ran into the error regarding wheel not building correctly.

Following another comment, changing the version number of PyYAML to 6.0.1 in requirements.txt fixed that issue.  I seem to be finished with the installation.

**Comment by AzorianMatt** [comment]
This is not a bug, but a deployment issue. While the documentation is almost certainly not up-to-date for FreeBSD deployments, the instructions for Debian / Ubuntu still seem pretty close if not usable as I have no issues running the project in my Ubuntu 22.04 environment.

I'm converting this to a discussion.
